Add worker indexing code path to RpcFunctionInvocationDispatcher

anandagopal6 opened the pull request:
<!-- Please provide all the information below.  -->

### Split up channel initialization into multiple functions in RpcFunctionInvocationDispatcher

resolves #7550 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

anandagopal6 committed:
Add worker indexing code path to RpcFunctionInvocationDispatcher.

This requires breaking down InitializeAsync because the second half of channel initialization can only occur after the worker has sent the function metadata.

anandagopal6 committed:
Update RpcFunctionInvocationDispatcherTests.cs

Setting an environment variable manually caused some tests to fail. Removing that for now.

anandagopal6 committed:
Update RpcFunctionInvocationDispatcher.cs

Changed a comment to be more clear

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -52,6 +52,9 @@ internal class RpcFunctionInvocationDispatcher : IFunctionInvocationDispatcher
         private TimeSpan _shutdownTimeout;
+        private bool _workerIndexing;
+        private ConcurrentBag<IRpcWorkerChannel> _channels = new ConcurrentBag<IRpcWorkerChannel>();
+        private List<FunctionMetadata> _rawMetadata = new List<FunctionMetadata>();
```
Should be IEnumerable

anandagopal6 replied:
Done!

soninaren replied:
Not sure if the change missed the commit.

anandagopal6 replied:
Think I might have missed this, just made the change though!

soninaren commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -201,5 +201,20 @@ public async Task<bool> RestartWorkerWithInvocationIdAsync(string invocationId)
+            throw new NotImplementedException();
+        }
+
+        public void FinishInitialization(List<FunctionMetadata> functions, CancellationToken cancellationToken = default)
```
IEnumerable

anandagopal6 replied:
Done!

soninaren commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -201,5 +201,20 @@ public async Task<bool> RestartWorkerWithInvocationIdAsync(string invocationId)
+            throw new NotImplementedException();
+        }
+
+        public Task<List<FunctionMetadata>> GetWorkerMetadata()
```
IEnumerable

anandagopal6 replied:
Done!

soninaren commented on src/WebJobs.Script/Workers/IFunctionInvocationDispatcher.cs:
```diff
@@ -24,5 +24,11 @@ public interface IFunctionInvocationDispatcher : IDisposable
+
+        void FinishInitialization(List<FunctionMetadata> functions, CancellationToken cancellationToken = default);
+
+        Task<List<FunctionMetadata>> GetWorkerMetadata();
```
IEnumerable

anandagopal6 replied:
Done!

soninaren commented on src/WebJobs.Script/Workers/IFunctionInvocationDispatcher.cs:
```diff
@@ -24,5 +24,11 @@ public interface IFunctionInvocationDispatcher : IDisposable
+
+        Task StartInitialization(CancellationToken cancellationToken = default);
+
+        void FinishInitialization(List<FunctionMetadata> functions, CancellationToken cancellationToken = default);
```
IEnumerable

anandagopal6 replied:
Done!

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -109,13 +112,25 @@ internal async Task InitializeJobhostLanguageWorkerChannelAsync()
+            }
+            else
+            {
+                var rpcWorkerChannel = _rpcWorkerChannelFactory.Create(_scriptOptions.RootScriptPath, _workerRuntime, _metricsLogger, attemptCount, _workerConfigs);
```
move out of if

anandagopal6 replied:
Done!

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -235,6 +260,120 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
+        public async Task StartInitialization(CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            List<FunctionMetadata> temp = new List<FunctionMetadata>();
```
move after the if

anandagopal6 replied:
Done!

soninaren reviewed (changes requested):
Adding first set of comments

anandagopal6 committed:
Removed most of the code duplication in RpcFunctionInvocationDispatcher, changed List instantiations to IEnumerable as requested, and merged StartInitialization into InitializeAsync.

To prevent any behavioral changes within InitializeAsync, the new merged changes into this function are behind a boolean that is set to false for now.

soninaren commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -201,5 +202,25 @@ public async Task<bool> RestartWorkerWithInvocationIdAsync(string invocationId)
+            throw new NotImplementedException();
+        }
+
+        public Task<(List<FunctionMetadata>, List<IRpcWorkerChannel>)> StartInitialization(CancellationToken cancellationToken = default)
```
don't think this is needed anymore.

anandagopal6 replied:
You're right, changed this in the latest commit!

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -110,12 +113,18 @@ internal async Task InitializeJobhostLanguageWorkerChannelAsync()
             _logger.LogDebug("Adding jobhost language worker channel for runtime: {language}. workerId:{id}", _workerRuntime, rpcWorkerChannel.Id);
-            rpcWorkerChannel.SendFunctionLoadRequests(_managedDependencyOptions.Value, _scriptOptions.FunctionTimeout);
-            SetFunctionDispatcherStateToInitializedAndLog();
+            _channels.Add(rpcWorkerChannel);
```
Nit: Can likely be moved next to _jobHostLanguageWorkerChannelManager.AddChannel

anandagopal6 replied:
done!

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -170,6 +185,10 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
             _functions = functions;
+
+            // feature flag and capability check will determine the value of _workerIndexing in a future worker indexing PR
+            _workerIndexing = false;
```
This should be set based on feature flag and capabilities fetch.

anandagopal6 replied:
Feature flags and capabilities are introduced in a future PR, so I just have this set to false for now and will update these later on.

anandagopal6 replied:
Just added the `EnableWorkerIndexing` feature flag in the latest commit, so for now, _workerIndexing is based on whether the feature flag is enabled or not.

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -230,8 +254,57 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
+            var channels = _channels.ToList();
+            if (channels.Count > 0)
+            {
+                return await channels[0].WorkerGetFunctionMetadata();
```
Can use channels.FirstOrDefault()?.WorkerGetFunctionMetadata()

anandagopal6 replied:
Good call, I'll change this to channels.First() because we already check to see if channels has at least one element in it

anandagopal6 committed:
Remove unnecessary function headers in HttpFunctionInvocationDispatcher

soninaren reviewed (changes requested):
Few more changes

anandagopal6 committed:
Changed some details brought up in PR review.

anandagopal6 committed:
Added "EnableWorkerIndexing" feature flag and made List<FunctionMetadata> variable types more generic with IEnumerable type.

soninaren commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -170,6 +186,17 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
             _functions = functions;
+
+            // feature flag and capability check will determine the value of _workerIndexing in a future worker indexing PR
+            if (FeatureFlags.IsEnabled(ScriptConstants.FeatureFlagEnableWorkerIndexing))
```
can just do `_workerIndexing = FeatureFlags.IsEnabled(ScriptConstants.FeatureFlagEnableWorkerIndexing)`

anandagopal6 replied:
so true.. done!

anandagopal6 commented:
/azp run

anandagopal6 committed:
Simplifying feature flag check to one line

pragnagopa commented:
cc @alrod / @AnatoliB  as FYI

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -170,6 +186,10 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
             _functions = functions;
+
+            // feature flag and capability check will determine the value of _workerIndexing in a future worker indexing PR
+            _workerIndexing = FeatureFlags.IsEnabled(ScriptConstants.FeatureFlagEnableWorkerIndexing);
```
This needs be checked ahead of time here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Host/ScriptHost.cs#L283

Host has logic that finds functions and indexes as noted above - this is needed for status end point that lists Functions Status / Errors - That code needs to run after worker sends FUnctions Metadata. 

Also, look at https://github.com/Azure/azure-functions-host/blob/128e7feaa8fb887fc23a5d42ad7cfc5d2bf4c3fc/src/WebJobs.Script/Host/ScriptHost.cs#L336 --> This collects indexing errors per function -- any errors that are cached by the worker and sent to host need to be reflected here.

anandagopal6 commented:
Closing this in favor of the larger PR that contains all Stein changes
